File: iOS/Spaces V2/Models/SpaceFile.swift
import Foundation

struct SpaceFolder: Identifiable, Hashable {
    let id: String
    let name: String
    let createdBy: String
    let createdAt: Date?

    var timestamp: String {
        guard let createdAt else { return "Recently" }
        return Self.timestampFormatter.localizedString(for: createdAt, relativeTo: Date())
    }

    private static let timestampFormatter: RelativeDateTimeFormatter = {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return formatter
    }()
}

struct SpaceFileItem: Identifiable, Hashable {
    let id: String
    let spaceID: String
    let name: String
    let mimeType: String
    let folderId: String?
    let storagePath: String
    let encryptionVersion: String
    let nonceBase64: String
    let uploadedBy: String
    let uploadedByName: String
    let fileExtension: String
    let createdAt: Date?
    let updatedAt: Date?
    let sizeBytes: Int64
    let deleted: Bool

    var timestamp: String {
        guard let createdAt else { return "Recently" }
        return Self.timestampFormatter.localizedString(for: createdAt, relativeTo: Date())
    }

    var sizeDescription: String {
        ByteCountFormatter.string(fromByteCount: sizeBytes, countStyle: .file)
    }

    var iconName: String {
        if mimeType.hasPrefix("image/") {
            return "photo"
        }
        if mimeType.hasPrefix("video/") {
            return "film"
        }
        if mimeType.hasPrefix("audio/") {
            return "waveform"
        }
        if mimeType.contains("pdf") {
            return "doc.richtext"
        }
        if mimeType.contains("zip") || mimeType.contains("compressed") {
            return "archivebox"
        }
        if mimeType.contains("json") || mimeType.contains("csv") || mimeType.hasPrefix("text/") {
            return "doc.text"
        }
        return "doc"
    }

    var typeDescription: String {
        if mimeType.hasPrefix("image/") {
            return "Image"
        }
        if mimeType.hasPrefix("video/") {
            return "Video"
        }
        if mimeType.hasPrefix("audio/") {
            return "Audio"
        }
        if mimeType.contains("pdf") {
            return "PDF"
        }
        if mimeType.contains("spreadsheet") || mimeType.contains("csv") {
            return "Spreadsheet"
        }
        if mimeType.contains("presentation") {
            return "Presentation"
        }
        if mimeType.contains("wordprocessingml") || mimeType.contains("msword") {
            return "Document"
        }
        if mimeType.contains("zip") || mimeType.contains("compressed") {
            return "Archive"
        }
        if mimeType.contains("json") {
            return "JSON"
        }
        if mimeType.hasPrefix("text/") {
            return "Text"
        }
        return "File"
    }

    var isImage: Bool {
        mimeType.hasPrefix("image/")
    }

    var isVideo: Bool {
        mimeType.hasPrefix("video/")
    }

    private static let timestampFormatter: RelativeDateTimeFormatter = {
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return formatter
    }()
}

File: Android V1.6/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceFile.kt
package com.arcinteractive.spaces.data.model

import java.util.Date

data class SpaceFolder(
    val id: String,
    val name: String,
    val createdBy: String,
    val createdAt: Date?
)

data class SpaceFileItem(
    val id: String,
    val spaceId: String,
    val name: String,
    val mimeType: String,
    val folderId: String?,
    val storagePath: String,
    val encryptionVersion: String,
    val nonceBase64: String,
    val uploadedBy: String,
    val uploadedByName: String,
    val fileExtension: String,
    val createdAt: Date?,
    val updatedAt: Date?,
    val sizeBytes: Long,
    val deleted: Boolean
) {
    val iconEmoji: String
        get() = when {
            mimeType.startsWith("image/") -> "\uD83D\uDDBC\uFE0F"
            mimeType.startsWith("video/") -> "\uD83C\uDFA5"
            mimeType.startsWith("audio/") -> "\uD83C\uDF99\uFE0F"
            mimeType.contains("pdf", ignoreCase = true) -> "\uD83D\uDCC4"
            mimeType.contains("zip", ignoreCase = true) || mimeType.contains("compressed", ignoreCase = true) -> "\uD83D\uDCE6"
            mimeType.contains("json", ignoreCase = true) || mimeType.contains("csv", ignoreCase = true) || mimeType.startsWith("text/") -> "\uD83D\uDCDD"
            else -> "\uD83D\uDCC1"
        }

    val sizeDescription: String
        get() = humanReadableByteCount(sizeBytes)

    val typeDescription: String
        get() = when {
            mimeType.startsWith("image/") -> "Image"
            mimeType.startsWith("video/") -> "Video"
            mimeType.startsWith("audio/") -> "Audio"
            mimeType.contains("pdf", ignoreCase = true) -> "PDF"
            mimeType.contains("spreadsheet", ignoreCase = true) || mimeType.contains("csv", ignoreCase = true) -> "Spreadsheet"
            mimeType.contains("presentation", ignoreCase = true) -> "Presentation"
            mimeType.contains("wordprocessingml", ignoreCase = true) || mimeType.contains("msword", ignoreCase = true) -> "Document"
            mimeType.contains("zip", ignoreCase = true) || mimeType.contains("compressed", ignoreCase = true) -> "Archive"
            mimeType.contains("json", ignoreCase = true) -> "JSON"
            mimeType.startsWith("text/") -> "Text"
            else -> "File"
        }

    val isImage: Boolean
        get() = mimeType.startsWith("image/")

    val isVideo: Boolean
        get() = mimeType.startsWith("video/")
}

private fun humanReadableByteCount(bytes: Long): String {
    if (bytes < 1024) return "$bytes B"
    val units = arrayOf("KB", "MB", "GB", "TB", "PB", "EB")
    var value = bytes.toDouble()
    var unitIndex = -1
    while (value >= 1024 && unitIndex < units.lastIndex) {
        value /= 1024
        unitIndex += 1
    }
    return String.format("%.1f %s", value, units[unitIndex])
}
